rfcs/2021-10-14-9811-vrl-performance.md

```diff
@@ -0,0 +1,374 @@
+
+Also, each node in the AST is boxed which causes extra pointer indirection.
+
+
+## Proposal
+
+### Implementation
+
+### Reducing allocations
+
+#### Use reference counting
+
+One of the data types that is cloned the most is `vrl::Value`. A lot of these
+clones are because VRL needs to maintain multiple references to a single value.
+In the Rust model, this can get highly awkward handling the lifetimes.
+
+If VRL used reference counting instead, the allocations can be avoided. Each
+clone would just increase a count into the underlying data store which is
+significantly cheaper. When the value goes out of scope the count is reduced.
+If the count reaches 0, the memory is freed.
+
+VRL needs to wrap the data in an `Rc`. This does mean pulling the data out of
+the original event, which is not wrapped in `Rc`.  Since we already clone the
+data at the start of the transform this is not going to create any additional
+cost to what we already have.
+
+Each field in the event can be pulled out lazily as it is used.
+
+At the end of the process any modified data can be reapplied to the event. This
+should be possible with a simple `mem::swap`. If the process is aborted, we
+don't do anything since the original event is still unmodified.
+
+Currently, the only point at which data is mutated is when updating the event.
+With this change mutation would occur when updating the VRL data store.
+
+Some changes will need to be made:
+
+1. Change `vrl::Value::Object` to wrap the Value in `Rc<RefCell<>>`:
+   `Object(BTreeMap<String, Rc<RefCell<Value>>>)`
+2. Change `vrl::Value::Array`  to wrap the Value in `Rc<RefCell<>>`:
+   `Array(Vec<Rc<RefCell<Value>>>)`
+3. Change the methods of the `Target` trait to use `Rc<RefCell<>>`:
+
+```rust
+    fn insert(&mut self, path: &LookupBuf, value: Rc<RefCell<Value>>) -> Result<(), String>;
+    fn get(&self, path: &LookupBuf) -> Result<Option<Rc<RefCell<Value>>>, String>;
+    fn remove(
+        &mut self,
+        path: &LookupBuf,
+        compact: bool,
+    ) -> Result<Option<Rc<RefCell<Value>>>, String>;
+```
+
+4. Change `VrlTarget`, to store the event as a `<Rc<RefCell<vrl_core::Value>>`
+   rather than a `Value`.  This does mean that when the remap transform runs
+   all the data from `Value` will need converting to `vrl_core::Value`. There
+   is a cost involved here. However it does mean that we can avoid the initial
+   clone that is often necessary, which should soak up most of the extra
+   expense.
+5. Update everything to use `Rc<RefCell<Value>>`. There is a risk here since
+   using `RefCell` moves the borrow checking to the runtime rather than compile
+   time. Affected areas of code will need to be very vigorously tested to avoid
+   runtime panics. Fortunately a lot of Vrl relies on creating new values
+   rather than mutating existing, so there isn't too much code that will be
+   affected.
+6. `Value` is no longer `Send + Sync`. There are some nodes in the AST that
```

Comment:
The ultimate reason Vrl needs `Send + Sync` is because it needs to store the `Program` in the transform and the `FunctionTransform` trait needs `Send + Sync`. Is there a way around that?

Reply:
Ah, sorry, I meant to say I was pro the removal of `Send + Sync` from `Value` and not remap generally.